akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala

```diff
@@ -269,6 +270,57 @@ object GraphStages {
   def terminationWatcher[T]: GraphStageWithMaterializedValue[FlowShape[T, T], Future[Done]] =
     TerminationWatcher.asInstanceOf[GraphStageWithMaterializedValue[FlowShape[T, T], Future[Done]]]
 
+  private class FlowMonitorImpl[T] extends FlowMonitor[T] {
+    private val stateRef = new AtomicReference[StreamState[T]](Initialized)
+    def setState(state: StreamState[T]): Unit = stateRef.set(state)
+
+    override def state = stateRef.get()
+  }
```

Comment:
it might be possible to save on allocation of a Received for every message if you store it in an AnyRef field instead of type T, and wrapping it only back on read if needed. Roughly:
```scala
get() match {
  case s: StreamState => s
  case elem => Received(elem)
}

```